Add missing return type annotation for named_modules()

yashwantbezawada opened the pull request:
## Summary

This PR adds the missing return type annotation `-> Iterator[tuple[str, Module]]` to the `named_modules()` method, bringing it in line with similar methods like `named_children()` which already have proper type annotations.

## Changes

1. **torch/nn/modules/module.py**: Added return type annotation `-> Iterator[tuple[str, "Module"]]` to `Module.named_modules()`
2. **torch/distributed/nn/api/remote_module.py**: Added return type annotation `-> Iterator[tuple[str, Module]]` to `_RemoteModule.named_modules()`

## Motivation

The `named_modules()` method was missing a return type annotation, which was causing issues for type checkers. Other similar methods like `named_children()` already have this annotation, so this PR ensures consistency across the codebase.

## Testing

- Python syntax validation passed
- Type annotations match the pattern used by `named_children()` and other iterator methods

Fixes #166905

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @lolpack @maggiemoss @ndmitchell @kinto0 @samwgoldman

yashwantbezawada committed:
Add missing return type annotation for named_modules()

Add return type annotation `-> Iterator[tuple[str, Module]]` to
`named_modules()` method in both `nn.Module` and `_RemoteModule`.

This brings `named_modules()` in line with other similar methods like
`named_children()` which already have proper type annotations.

Fixes #166905

yashwantbezawada commented:
@pytorchbot label "topic: not user facing"

yashwantbezawada committed:
Fix type error in param_fetch.py due to named_modules() annotation

Add type assertion to clarify that node.target is a string when
node.op is 'call_module'. This resolves the pyrefly lint error that
was exposed by adding the return type annotation to named_modules().

yashwantbezawada commented on torch/fx/passes/param_fetch.py:
```diff
@@ -89,6 +89,8 @@ def lift_lowering_attrs_to_nodes(fx_module: GraphModule) -> None:
     for node in fx_module.graph.nodes:
         if node.op == "call_module":
+            # When op is "call_module", target is always a string (module name)
+            assert isinstance(node.target, str)
```
yeah that's fair — the reason for it is that adding the return type to named_modules() makes dict(fx_module.named_modules()) a dict[str, Module], and node.target is typed as Union[Callable, str], so the type checker flags submodules[node.target] without narrowing. the same assert isinstance(node.target, str) pattern shows up a bunch of times across torch/fx/ (graph.py, graph_module.py, split_module.py, etc.) for exactly this reason. happy to swap it for a `# pyrefly: ignore` comment instead if you'd prefer though

Skylion007 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -438,7 +438,7 @@ def named_modules(
         prefix: str = "",
         remove_duplicate: bool = True,
-    ):
+    ) -> Iterator[tuple[str, Module]]:
```
Shouldn't this be return type never or does the typing system throw a fit due to the base class here?

yashwantbezawada committed:
Add type: ignore[return] and pyrefly: ignore to named_modules()

Match the pattern used by named_parameters() and named_buffers()
in _RemoteModule, since the method always raises via
_raise_not_supported() which is typed as -> None.
